### ISSUE TITLE
Relative Color Syntax (RCS)

### ISSUE THREAD

**Issue by mysteryDate** [w3ctag/design-reviews#894]
こんにちは TAG-さん!

I'm requesting a TAG review of Relative Color Syntax (RCS).

The new relative color syntax extends modern color syntax to allow existing colors to be modified using the color functions: if an origin color is specified, then each color channel can either be directly specified, or taken from the origin color (and possibly modified with math functions). For example:

`lab(from magenta calc(l/2) a b)`

Would create a `lab` color that is equivalent to magenta with half of the luminance.

  - Explainer : https://www.stefanjudis.com/notes/new-in-css-relative-colors/
  - Specification URL: https://www.w3.org/TR/css-color-5/#relative-colors
  - Tests:
    - https://github.com/web-platform-tests/wpt/blob/master/css/css-color/parsing/color-computed-relative-color.html
    - https://github.com/web-platform-tests/wpt/blob/master/css/css-color/parsing/color-valid-relative-color.html
    - https://github.com/web-platform-tests/wpt/blob/master/css/css-color/parsing/color-invalid-relative-color.html
    - https://github.com/web-platform-tests/wpt/blob/master/css/css-color/parsing/relative-color-out-of-gamut.html
  - Primary contacts (and their relationship to the specification):
      - Chris Lilley [svgeesus](https://github.com/svgeesus) W3C 
      - Adam Argyle [argyleink](https://github.com/argyleink) Chromium 
  - Organization(s)/project(s) driving the specification: W3C, Webkit, Chromium
  - Key pieces of existing multi-stakeholder review or discussion of this specification: Original proposal here https://github.com/w3c/csswg-drafts/issues/3187
  - External status/issue trackers for this specification (publicly visible, e.g. Chrome Status): https://chromestatus.com/feature/5205844613922816

Further details:

  - [x] I have reviewed the TAG's [Web Platform Design Principles](https://www.w3.org/TR/design-principles/)
  - The group where the work on this specification is currently being done: [csswg-drafts](https://github.com/w3c/csswg-drafts)
  - Major unresolved issues with or opposition to this specification: None
  - This work is being funded by: Google

**Comment by LeaVerou** [comment]
> * Primary contacts (and their relationship to the specification):
>   
>   * Chris Lilley [svgeesus](https://github.com/svgeesus) W3C
>   * Adam Argyle [argyleink](https://github.com/argyleink) Chromium

Hi there,

Curious why list only 2 of the 4 editors of the specification as primary contacts? Especially since the editor who [actually proposed the feature](https://github.com/w3c/csswg-drafts/issues/3187) is among the 2 that were left out.

**Comment by mysteryDate** [comment]
> Curious why list only 2 of the 4 editors of the specification as primary contacts? Especially since the editor who [actually proposed the feature](https://github.com/w3c/csswg-drafts/issues/3187) is among the 2 that were left out.

I was trying to keep it brief and just listed the two authors with whom I'd actively been in contact. Would you like me to add the other two to the original issue post?

**Comment by ZoeBijl** [comment]
Is there an explainer for this that’s not a blog post?

**Comment by myakura** [comment]
> Is there an explainer for this that’s not a blog post?

unfortunately not.
https://groups.google.com/a/chromium.org/g/blink-dev/c/KeKBTjKEdaU

**Comment by mysteryDate** [comment]
> Is there an explainer for this that’s not a blog post?

@LeaVerou Was there an explainer that you all were working from when the spec was first written? This is not the first time that someone has wanted something more official. I could write one now, but that would feel a little post-hoc and I feel like I am far from the most knowledgable person on the subject.

**Comment by LeaVerou** [comment]
> > Is there an explainer for this that’s not a blog post?
> 
> @LeaVerou Was there an explainer that you all were working from when the spec was first written? This is not the first time that someone has wanted something more official. I could write one now, but that would feel a little post-hoc and I feel like I am far from the most knowledgable person on the subject.

No, but it should be pretty easy to put one together with the info from [the original proposal](https://github.com/w3c/csswg-drafts/issues/3187) (which includes motivation, use cases, etc) and [the updated proposal](https://github.com/w3c/csswg-drafts/issues/3187#issuecomment-499126198) which converged on the sytnax that is currently in the spec.

I believe the differences between the current syntax and the (updated) proposal are:
- Using a gamut-restricted color function does not clip or gamut map anymore
- RCS for `color()` is now much more [fleshed out](https://drafts.csswg.org/css-color-5/#relative-color-function) than it was there.
- The return types for each keyword are now more fleshed out, defaulting to `<number>` when that is allowed for the corresponding component.

@svgeesus, am I forgetting any?

**Comment by miketaylr** [comment]
The above comment and its links is as useful as an explainer; it doesn't seem worth the effort to write a new one (but if someone is really inclined to, more power to them).

**Comment by LeaVerou** [comment]
> The above comment and its links is as useful as an explainer; it doesn't seem worth the effort to write a new one (but if someone is really inclined to, more power to them).

I disagree, the whole point of explainers is that they are easy to digest, piecing the info together from several links and cross-referencing with the spec is not work that whoever is reviewing the explainer should need to do.

I would love to do the work of putting an explainer together, but I can't commit to a specific timeframe right now, and I don't want to become the bottleneck here.

**Comment by svgeesus** [comment]
The other useful item that an explainer would provide is evidence of implementer support, so:

- [Is relative color syntax ready to ship?](https://github.com/w3c/csswg-drafts/issues/7978) RCS cleared to ship pre-CR by CSS WG
- [CanIUse relative colors](https://caniuse.com/css-relative-colors) shipping in Safari 16.4 onwards
- [RFP from Mozilla on RCS](https://github.com/mozilla/standards-positions/issues/841) with no answer yet
- [WPT results for RCS](https://wpt.fyi/results/css/css-color/parsing?label=master&label=experimental&aligned&q=relative) implemented in Chromium and WebKit

**Comment by svgeesus** [comment]
Oh btw @mysteryDate 

> Would create a `lab` color that is equivalent to magenta with half of the luminance

Half the lightness.

**Comment by svgeesus** [comment]
Ok here is a somewhat minimal  [RCS Explainer](https://github.com/w3c/csswg-drafts/blob/main/css-color-5/RCS-Explainer.md)

**Comment by mysteryDate** [comment]
Exellent, thank you! I have updated the original comment to add the two missing authors, link to the new explainer and swap "luminance" for "lightness."

**Comment by svgeesus** [comment]
I [added tests for RCS with `currentColor`](https://wpt.fyi/results/css/css-color?label=master&label=experimental&aligned&q=relative) which currently no-one passes, but the spec [clearly describes](https://drafts.csswg.org/css-color-5/#resolving-rcs) what should happen so hopefully the implementations will converge to interop on that aspect.

**Comment by svgeesus** [comment]
Mozilla position on RCS is [positive](https://mozilla.github.io/standards-positions/#css-relative-color-syntax) with implementation starting soon

**Comment by rhiaro** [comment]
Thanks for the explainer updates, no concerns from us. :sparkles:
